### PR TITLE
Remove rules duplicating method signature info

### PR DESCRIPTION
**Pull request by Vinai**
Remove rules that duplicate information in method signatures.

## Purpose of this pull request

This pull request (PR) removes the rules enforcing useless phpdoc annotations.
These useless annotations are generally not liked and add no information beyond what already is supplied by the class, interface or method name and the method signature.

## Affected DevDocs pages

- https://devdocs.magento.com/guides/v2.2/coding-standards/docblock-standard-general.html
- https://devdocs.magento.com/guides/v2.3/coding-standards/docblock-standard-general.html

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

**Comment by Vinai** [comment]
Please hold this PR until after it was discussed in an architetcure meeting.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 1]
```diff
@@ -326,27 +328,29 @@ class Profiler
...
-  If there is no such operator, the `@return` tag must have `void` as the return value.
+* The declaration of all arguments (if any) using `@param` tag, unless the argument type is indicated in the method signature.
+  All `@param` annotations must include the appropriate argument type.
+  If any argument requires a `@param` annotation, all arguments must be listed (all or none).
+* The declaration of the return type using the `@return` tag should only be added if the method return type signature
+  does not supply all necessary information.
+  If the method does not `return` a value, the method type signature should be `void`, e.g. `function example(): void {}`.
 * Declaration of possibly thrown exception using `@throws` tag, if the actual body of function triggers throwing an exception.
-  All occurrences of `@throws` in a DocBlock must be after `@param` and `@return` (if any).
+  All occurrences of `@throws` in a DocBlock must be after any `@param` and `@return` annotations (if any).
```
Comment (joshuaadickerson) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Remove the last "(if any)" or make it "must be after all".

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 2]
```diff
@@ -539,39 +546,27 @@ Documentation is organized with the use of namespaces.
...
 
-Whenever possible the `@inheritdoc` tag MUST be used for children to avoid duplication of doc blocks.
-
-Even Though PHPDocumentor understands inheritance and uses the parent doc block by default (without `@inheritdoc` tag specified), including the tag helps ensure that the doc block is not missed at all.
-
-Rules for usage of the tag:
-
-* Use `@inheritdoc` (notice no braces around) to indicate that the entire doc block should be inherited from the parent method.
-* Use the inline `{@inheritdoc}` tag (with braces around) in long descriptions to reuse the parent's long description. The tagged method MUST have its own short description.
+The `@inheritdoc` tag MUST NOT be used.
```
Comment (joshuaadickerson) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: This should be `SHOULD NOT` as the tag is supposed to be used for inline documentation. You may want to document the parent as _part_ of the child. Say for instance your child method just throws a bunch of exceptions. You'd use `@inheritdoc` and then `@throws` after it.

Comment (buskamuza) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Throwing new exception may be not the best example as it may break clients that depend on the parent. But adding long description that explains purpose of the method may require using `@inheritdoc`.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 1 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Done.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 2 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: done.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 3]
```diff
@@ -257,7 +257,9 @@ require_once __DIR__ . '/../../functions.php';
 ### Classes and interfaces
 {:#classes-interfaces}
 
-Classes and interfaces must have a short description that is a human-understandable intention of the class.
+Classes and interfaces should have a short description that is a human-understandable intention of the class.
+If the short description adds no additional information beyond what the type name already supplies, the
+short description should be ommitted.
```
Comment (DrewML) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: s/ommitted/omitted

**Comment by orlangur** [comment]
@buskamuza this PR is about removing ALL THE useless stuff.

> I'd suggest to allow `@inheritdoc`

Please provide some reasoning.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 4]
```diff
@@ -257,7 +257,9 @@ require_once __DIR__ . '/../../functions.php';
 ### Classes and interfaces
 {:#classes-interfaces}
 
-Classes and interfaces must have a short description that is a human-understandable intention of the class.
+Classes and interfaces should have a short description that is a human-understandable intention of the class.
+If the short description adds no additional information beyond what the type name already supplies, the
+short description should be omitted.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: MUST be omitted

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 5]
```diff
@@ -326,27 +328,29 @@ class Profiler
 ### Functions and methods
 {:#functions-methods}
 
-Functions and methods must have:
+In general, typed method signature should be preferred over PHPDoc annotations when possible.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: SHOULD -> MUST
when -> whenever

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 6]
```diff
@@ -326,27 +328,29 @@ class Profiler
...
 
-* Short description
-* Long description that explains the motivation behind the implementation.
+Functions and methods should have:
+
+* A short description in case it adds meaningful information beyond the method name.
+* If the purpose of the method is not obvious, a long description that explains the motivation behind the implementation.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Some sentence to involve "describe why method is implemented and not how", please

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 7]
```diff
@@ -326,27 +328,29 @@ class Profiler
...
    * For non-obvious implementations where the implementation logic is complicated or does not correspond to the Technical Vision or other known best practices, include an explanation in the doc block's description.
      An implementation is non-obvious if another developer has questions about it.
 
-* Declaration of all arguments (if any) using `@param` tag.
-  Appropriate argument type must be specified.
-* Declaration of return type using `@return` tag.
-  If there is no such operator, the `@return` tag must have `void` as the return value.
+* The declaration of all arguments (if any) using `@param` tag, unless the argument type is indicated in the method signature.
+  All `@param` annotations must include the appropriate argument type.
+  If any argument requires a `@param` annotation, all arguments must be listed (all or none).
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Disagree. Let's specify `@param` only for necessary arguments, when it is more accurate than type hinting.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 8]
```diff
@@ -326,27 +328,29 @@ class Profiler
...
+* The declaration of the return type using the `@return` tag should only be added if the method return type signature
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: should -> must

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 9]
```diff
@@ -326,27 +328,29 @@ class Profiler
...
+  does not supply all necessary information.
+  If the method does not `return` a value, the method type signature should be `void`, e.g. `function example(): void {}`.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Something like "does not have return statement" would be more clear. should -> must

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 10]
```diff
@@ -326,27 +328,29 @@ class Profiler
...
 * Declaration of possibly thrown exception using `@throws` tag, if the actual body of function triggers throwing an exception.
-  All occurrences of `@throws` in a DocBlock must be after `@param` and `@return` (if any).
+  All occurrences of `@throws` in a DocBlock must be after any `@param` and `@return` annotations.
 
 **Exceptions to these rules:**
 
-* Constructors may not have short and/or long description
-
 * Testing methods in Unit tests may not have doc blocks if the test's method name follows the convention (test<MethodName>)
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: `test<MethodNameInWhichCircumstances>` convention must be mandatory, thus doc block must be not allowed in most cases

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 11]
```diff
@@ -492,21 +496,24 @@ public function deleteDirectory($path)
 #### @return tag
 {:#return}
 
-If there is no explicit return statement in a method or function, a `@return void` should be used in the documentation.
+In general method return type signatures should be prefered over `@return` type annotations.
+If that is not possible due to ambiguous return types, the `@return` type annotation must be used.
+
+If there is no explicit return statement in a method or function, a `void` return type should be declarted in the method signature.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Hm, this one sounds better, should -> must and please remove the part "If the method does not `return` a value, the method type signature should be `void`, e.g. `function example(): void {}`."

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 12]
```diff
@@ -492,21 +496,24 @@ public function deleteDirectory($path)
...
 
-If the method returns itself, `return $this` should be used.
+If the method returns itself, `@return $this` should be used, unless the return type is specified in the method signature.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Should -> must. Looks like such annotation will remain for the sake of BC only.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 13]
```diff
@@ -539,39 +546,29 @@ Documentation is organized with the use of namespaces.
...
 
-Whenever possible the `@inheritdoc` tag MUST be used for children to avoid duplication of doc blocks.
-
-Even Though PHPDocumentor understands inheritance and uses the parent doc block by default (without `@inheritdoc` tag specified), including the tag helps ensure that the doc block is not missed at all.
-
-Rules for usage of the tag:
-
-* Use `@inheritdoc` (notice no braces around) to indicate that the entire doc block should be inherited from the parent method.
-* Use the inline `{@inheritdoc}` tag (with braces around) in long descriptions to reuse the parent's long description. The tagged method MUST have its own short description.
+The `@inheritdoc` tag SHOULD NOT be used.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: MUST NOT

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 14]
```diff
@@ -539,39 +546,29 @@ Documentation is organized with the use of namespaces.
...
+If a child class method requires a long description to explain its purpose, it may use `@inheritdoc` to indicate the new description is intended as an addition to the parent method description.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: We may allow `{@inheritdoc}` inline version only but I believe it would be better to forbid completely.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 15]
```diff
@@ -539,39 +546,29 @@ Documentation is organized with the use of namespaces.
...
+    public function getVal(): int;
 
     /**
-     * Set value
-     *
      * Sets 0 in case a non-integer value is passed
-     *
-     * @param int $value
+     * 
+     * @param mixed $value
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: I believe `mixed` must be avoided in favor of specific types (even if there are 3-4 of them).

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 16]
```diff
@@ -582,19 +579,16 @@ interface MutableInterface
...
     }
-
+    
     /**
-     * Set value
-     *
-     * Sets 0 in case the value is bigger than max allowed value. {@inheritdoc}
+     * Sets 0 in case a non-integer value is passed
+     * 
+     * @param mixed $value
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Same here for `mixed`

**Comment by Vinai** [comment]
@orlangur She did - I added that to the PR:

> The `@inheritdoc` tag SHOULD NOT be used.  
> If a child class method requires a long description to explain its purpose, it may use `@inheritdoc` to indicate the new description is intended as an addition to the parent method description.  
> In general such method overrides are a code smell and should be used as an incentive to make the code more self-documenting if possible.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 7 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: In my experience that leads to less readable code.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 9 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: A return statement without a return value would not be covered by that expression.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 15 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: But what if they are an infinite number of possible types, for example anything that can be cast to a string?

**Comment by orlangur** [comment]
@Vinai also, I'll double check regarding v2.1 and v2.2 in terms of used sniffs but my gut feeling says that we should only revise v2.3 here (I would be able to apply automated PHPDoc change for all the three release lines though).

**Comment by Vinai** [comment]
@orlangur I used the 2.1 version of the page because 2.2 and 2.3 just referenced it.
Should I try to copy the page contents of 2.1 into 2.3 so it no longer is a reference, and make the changes there, instead?

**Comment by orlangur** [comment]
Ok, let me check automated checks in 2.1/2.2 first and then will ask somebody from DevDocs team.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 15 above)
Comment (orlangur) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: True :) "should be avoided whenever possible" then. Still, I would get rid all of `mixed` from code samples in doc though.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 13 above)
Comment (orlangur) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Aha, let me reach Olga in Slack.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 7 above)
Comment (orlangur) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Some examples? I don't think `@param` annotations are for humans at all, they are to give maximum accurate information for IDE. Besides constructor, any method don't contain more than 4-5 parameters usually, so, I believe such block with partial types for arguments missing real type hinting will be readable anyway.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 15 above)
Comment (orlangur) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: @Vinai,

> But what if they are an infinite number of possible types

Changed my mind. Some interface or pseudo-type must be used. If there are too many different classes possible, just use `object`.

So, I would forbid `mixed` completely but this and few other my items can be added to separate PR so that this one only concentrates on stripping redundant declarations.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 7 above)
Comment (joshuaadickerson) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: I know default PhpStorm will complain about partial doc blocks. We could change this, but a lot of developers will not understand it and it will lead to inconsistencies. Not sure how IDEs or PhpDoc will handle the order of the arguments then.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 14 above)
Comment (joshuaadickerson) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Agreed. I don't think I've ever seen someone using inline inheritdoc without it looking weird in the documentation.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 17]
```diff
@@ -364,7 +368,7 @@ Functions and methods must have:
     For example: `@return Config|null`.
     The DockBlock needs to explain what situations return `null`.
 
-    Another example: `@param FileInterface | null`.
+    Another example: `@param FileInterface|null`.
```
Comment (joshuaadickerson) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Any reason you didn't opt for the nullable operator? `@param ?FileInterface`

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 7 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: I don't think there is an absolute truth to be reached in this regard (as is often the case in regards to "understandability". Who can decide this matter?

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 14 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: What's the current state in regards to `@inheritdoc`? I'm all for getting rid of it, but if @buskamuza wants to allow it I'm okay with that, too, as long as it's discouraged to use it situations where it is only noise.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 15 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: I got rid of `mixed` annotations in the examples.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 17 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: I find it strange to use nullable on `@param` annotations. Is that a good idea? I honestly don't know. So far I've never encountered that in annotations. Happy to change it if you think it's good and PHPStorm understands it.

**Comment by Vinai** [comment]
@orlangur I've applied most changes you requested in https://github.com/magento/devdocs/pull/4425#pullrequestreview-234103595

However:

> let's make no unnecessary alignment mandatory

and

> Let's make `/** @var Profiler */` the only form allowed.

I'm against those changes - they give me the feeling as if I'm beeing micro managed by someone oppressive.
As long as the style is consistent within a file I think it's better to allow either format. I've been discouraged from opening PRs or following through with PRs in the past by overzealous coding standard rules (e.g. in the phpunit/mink project).

**Comment by orlangur** [comment]
@Vinai thanks, no problem, to me those make sense if they are supported by auto-formatting tooling. Will try to address them separately.

Asked Olga in Slack regarding inheritdoc and will double-check the overall document in next few days. Huge thanks on your efforts to aggregate all the pain points related to redundancy 👍

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 18]
```diff
@@ -326,32 +328,34 @@ class Profiler
...
 
 **Exceptions to these rules:**
 
-* Constructors may not have short and/or long description
-
-* Testing methods in Unit tests may not have doc blocks if the test's method name follows the convention (test<MethodName>)
+* Testing methods in Unit tests may have doc blocks to describe the purpose of the test, for example referencing github issues.
 
-   * If the test does not follow the convention, it should have a doc block describing the covered methods
+* Test emthod annotations may include data providers and other testing annotations.
```
Comment (lenaorobei) [inline guides/v2.1/coding-standards/docblock-standard-general.md]:
``` suggestion
* Test method annotations may include data providers and other testing annotations.
```

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 7 above)
Comment (lenaorobei) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Can we also add a rule that `@param` annotations should be in the same order as method arguments?

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 18 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: thanks, fixed!

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 7 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: @lenaorobei done

**Comment by Vinai** [comment]
@buskamuza Your requested change to allow `@inheritdoc` is in. Are there other changes you would like to request?

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 7 above)
Comment (orlangur) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: @Vinai,

> In my experience incomplete lists of params in the annotations lead to less readable code.

Do you have some examples? There are not a lot of situations when we will need annotation for parameter at all (only more accurate type like `array` -> `Product[]` and such). Methods beside constructor usually won't contain more than 3-4 arguments.

Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: I don't have any examples at hand, since I changed those to contain all parameters instead.
But here is a constructed example:

Single `@param` annotation:
```php
/**
 * @param SectionSourceInterface[] $sectionSourceMap
 */
public function __construct(
    ObjectManagerInterface $objectManager,
    SectionIdentifier $identifier,
    array $sectionSourceMap = []
) {
    $this->objectManager = $objectManager;
    $this->identifier = $identifier;
    $this->sectionSourceMap = $sectionSourceMap;
}
```

A complete list of parameters annotations:
```php
/**
 * @param ObjectManagerInterface $objectManager
 * @param SectionIdentifier $identifier
 * @param SectionSourceInterface[] $sectionSourceMap
 */
public function __construct(
    ObjectManagerInterface $objectManager,
    SectionIdentifier $identifier,
    array $sectionSourceMap = []
) {
    $this->objectManager = $objectManager;
    $this->identifier = $identifier;
    $this->sectionSourceMap = $sectionSourceMap;
}
```

**Comment by Vinai** [comment]
@orlangur I'm happy to do whatever in regards to `@inheritdoc` - please tell me what to do and I'll adjust the PR accordingly.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 14 above)
Comment (buskamuza) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: I agree with keeping just "The `@inheritdoc` tag SHOULD NOT be used." if it's ok for everyone.
I don't see why it should be "MUST NOT", as it doesn't break anything. But I'm ok with discouraging.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 19]
```diff
@@ -425,7 +430,7 @@ In general, use the `@throws` tag when the code uses *throw*:
  *
  * @param  string $elementId
  * @param  string $attribute
- * @param  mixed  $value
+ * @param  int|string|float|bool|object|null  $value
```
Comment (paliarush) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Why is this change needed? It is not exactly equivalent transformation since `mixed` is broader.

Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: This was a request by @orlangur - I personally think `mixed` is fine, but don't feel strongly either way.

Comment (orlangur) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: @paliarush do you mean it can be `resource`? Generally `mixed` does not smell well.

**Comment by orlangur** [comment]
@Vinai @buskamuza in which case we need `@inheritdoc`? It does not bring any value and complicates review process.

@Vinai another thing we discussed earlier - could you provide example when partial parameters documentation may be confusing? We will need parameter type in PHPDoc only in rare cases, like a typified array `string[]` or smth.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 20]
```diff
@@ -326,32 +328,35 @@ class Profiler
...
      An implementation is non-obvious if another developer has questions about it.
 
-* Declaration of all arguments (if any) using `@param` tag.
-  Appropriate argument type must be specified.
-* Declaration of return type using `@return` tag.
-  If there is no such operator, the `@return` tag must have `void` as the return value.
+* The declaration of all arguments (if any) using `@param` tag, unless the argument type is indicated in the method signature.
+  All `@param` annotations must include the appropriate argument type.
+  If any argument requires a `@param` annotation, all arguments must be listed (all or none).
+* The `@param` annotations must be in the same order as the method arguments.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: I would combine it with "all arguments must be listed" item

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 21]
```diff
@@ -326,32 +328,35 @@ class Profiler
...
+* The declaration of the return type using the `@return` tag must only be added if the method return type signature
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: It is unclear whether `@return void` is still mandatory or not?

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 19 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Besides `resoruce`, the `@mixed` annotation includes any future types, too.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 22]
```diff
@@ -326,32 +328,35 @@ class Profiler
...
+  does not supply all necessary information (see below for more information on return types).
 * Declaration of possibly thrown exception using `@throws` tag, if the actual body of function triggers throwing an exception.
-  All occurrences of `@throws` in a DocBlock must be after `@param` and `@return` (if any).
+  All occurrences of `@throws` in a DocBlock must be after any `@param` and `@return` annotations.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: "any" may be omitted here at all.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 23]
```diff
@@ -326,32 +328,35 @@ class Profiler
...
 
 **Exceptions to these rules:**
 
-* Constructors may not have short and/or long description
-
-* Testing methods in Unit tests may not have doc blocks if the test's method name follows the convention (test<MethodName>)
+* Testing methods in Unit tests may have doc blocks to describe the purpose of the test, for example referencing github issues.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: Honestly, I like previous version more. My understanding is that we should insist  descriptive enough `test<MethodName>` method name without any PHPDoc is preferred.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 24]
```diff
@@ -326,32 +328,35 @@ class Profiler
...
 
-   * If the test does not follow the convention, it should have a doc block describing the covered methods
+* Test method annotations may include data providers and other testing annotations.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: This can be removed I believe.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 25]
```diff
@@ -326,32 +328,35 @@ class Profiler
...
 
-   * Non-testing methods should have a doc block with description. It includes data providers and any helper methods
+* Non-testing methods may have a doc block with description according to the rules above.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: This is not needed aswell: anything which is not forbidden is allowed.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 26]
```diff
@@ -492,21 +497,24 @@ public function deleteDirectory($path)
 #### @return tag
 {:#return}
 
-If there is no explicit return statement in a method or function, a `@return void` should be used in the documentation.
+In general method return type signatures should be prefered over `@return` type annotations.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: I would say

```
Return type signature MUST be used whenever possible.
`@return` annotation MUST be omitted when it brings no new information compared to type signature.
```

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 27]
```diff
@@ -492,21 +497,24 @@ public function deleteDirectory($path)
...
+If that is not possible due to ambiguous return types, the `@return` type annotation must be used.
+
+If there is no explicit return statement in a method or function or a return statement without a value, a `void` return type must be declared in the method signature.
 
-If the method returns itself, `return $this` should be used.
+If the method returns itself, `@return $this` must be used, unless the return type is specified in the method signature.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: So, what to use for `function FooBar(): self` ? I think no `@return` needed in this case.

When there is not type hint, I'm a bit concerned whether we need explicit `@return $this` or PhpStorm may resolve correct type anyway.

File: guides/v2.1/coding-standards/docblock-standard-general.md [hunk 28]
```diff
@@ -492,21 +497,24 @@ public function deleteDirectory($path)
...
 
 ### Constants
 {:#constants}
 
-Constants must have short description.
+Constants must have short description if they add information beyond what the constant name supplies.
```
Comment (orlangur) [inline guides/v2.1/coding-standards/docblock-standard-general.md]: It's better like in another section: `MUST be omitted when it brings no additional information`

**Comment by Vinai** [comment]
@orlangur Regarding partial vs. complete parameter annotations: I believe it boils down to familiarity which scenaerio is more or less confusing. We will not reach an agreement with examples here.

The use cases that require type annotations beyond what PHP is capable of typing are:

* typed array value annotations (like you said), e.g. `string[]` 
* quasi union types that have no common super type, e.g. `Foo|Bar`
* multiple scalar types, e.g. `string|int`

Just saying, because in my experience they aren't such a rare ocurance...

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 20 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: done

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 21 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Does the sentence exclude `void`in some way? To me it seems clear.
Should `void` be made explicit?
e.g: If a method return type signature is `void` then the `@return void` annotation must be omitted.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 22 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: done

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 23 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: A test name with a class name often is not sufficient. Since tests are documentation, I think it's good to be more permissive than the previous version of the sentence.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 24 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Because it is implicit?

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 25 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: removed

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 26 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Restricting types make code harder to evolve over time. I believe it is better to sometimes NOT specify a return type until a method signature is very stable. This more permissive statement allows for this.

**Comment by orlangur** [comment]
> Just saying, because in my experience they aren't such a rare occurrence...

@Vinai sure, but it won't happen for too many parameters of the same method. And such accurate type specification is more for IDE type resolution than human reading.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 27 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: Good point. I'll expand the section accordingly.

File: guides/v2.1/coding-standards/docblock-standard-general.md (hunk 28 above)
Comment (Vinai) [reply guides/v2.1/coding-standards/docblock-standard-general.md]: done

**Comment by Vinai** [comment]
Moved all changes to `guides/v2.2/coding-standards/docblock-standard-general.md` instead of the `2.1` file.

**Comment by dobooth** [comment]
running tests

**Comment by orlangur** [comment]
@Vinai glad to finally see it in, you rock 🎉 Special thanks for squashing prior to merge 😉 

Let's start the biggest cleanup ever made for core codebase.
